File: 2024/day02_Red-Nosed_Reports/main.py
import sys


def isSafe(report: list[int]) -> bool:
    sign = -1 if report[0] < report[1] else 1
    for i in range(len(report)-1):
        diff = (report[i] - report[i+1]) * sign
        if diff < 1 or diff > 3:
            return False
    return True

def isSafe2(report: list[int]) -> bool:
    sign = -1 if report[0] < report[1] else 1
    for i in range(len(report)-1):
        diff = (report[i] - report[i+1]) * sign
        if diff < 1 or diff > 3:
            return isSafe(report[1:]) or isSafe(report[:i] + report[i+1:]) or isSafe(report[:i+1] + report[i+2:])
    return True

def part1(file: list[str]):
    reports = map(lambda line: list(map(int, line.split())), file)

    res = len(list(filter(isSafe, reports)))
    print(res)

def part2(file: list[str]):
    reports = map(lambda line: list(map(int, line.split())), file)

    res = len(list(filter(isSafe2, reports)))
    print(res)


with open(sys.argv[1]+".txt", "r") as file:
    lines = [line.rstrip("\n") for line in file]

part1(lines)
part2(lines)

File: 2024/day02_Red-Nosed_Reports/src/main.rs
use std::io::stdin;

fn is_safe(report: &Vec<i32>) -> bool {
    let sign = if report[0] < report[1] { -1 } else { 1 };
    for i in 0..report.len()-1 {
        let diff = (report[i] - report[i+1]) * sign;
        if diff < 1 || diff > 3 {
            return false;
        }
    }
    return true;
}

fn part1(reports: &Vec<Vec<i32>>) {
    let res = reports
        .iter()
        .filter(|report| is_safe(report))
        .count();
    println!("{}", res);
}

fn main() {
    let mut reports = Vec::new();
    loop {
        let mut line = String::new();
        match stdin().read_line(&mut line) {
            Ok(0) => break,
            Ok(_) => reports.push(line
                .trim()
                .split_whitespace()
                .map(|x| x.parse().unwrap())
                .collect()
            ),
            Err(_) => break,
        }
    }
        

    part1(&reports);
    //part2(&file);
}
